### PR TITLE
Add type cast for number and fix select when not use with string

### PR DESCRIPTION
**Pull request by Zen3515**
This pull request is more of an enhancement than fix.

## 1. Enhance Number platform by introduce type casting and support step

- Type casting enable it to works with int float and string rather than just float.
- Step: just normal step, may be increase step by 0.5, 10 rather than 1

```yaml
      - platform: number
        friendly_name: Power Level
        id: 106
        min_value: 1 # Optional
        max_value: 9
        step_value: 1 # Optional
        dp_data_type: int # Optional default to float
```

## 2. Fix error when use select platform on non-string DPs

In the example below show the usecase where the select platform is used on the DPs that expect an Int

```yaml
      - platform: select
        friendly_name: Power On State
        id: 38
        select_options: 0;1;2
        select_options_friendly: Off;On;Remember last state
```

### CODE REVIEW

**Comment by Zen3515** [comment]
@rospogrigio Hi, could you approve this pull request?
Please let me know if there are any changes required.

**Comment by iMiMx** [comment]
Thanks for this - it seems to have partially fixed an issue I was having.

I have a dehumidifier, the humidity level settings are 15,35,40...... 80 - 15 is constant, then 35 -> 80 are actual humidity levels.  

I managed to map all the various other DPs, but this one would not work until I used your modifications (select would show 'unavailable' and number just wouldn't do anything).

With the changes to number and the ability to specify 'int' I can now create a slider (with 5 value step) that controls the humidity level.  

But, I think 'select' would be a better option (so that I can skip 20, 25, 30 values), whilst using this shows the values in HA, if I try to change them using the select entity the values do not making it to the device.

... long story short, I'm wondering if the CONF_DP_DTYPE = "dp_data_type" added to the selection option might also be valuable and resolve the problem?  I will try hacking around with your changes to see.

**Comment by Zen3515** [comment]
> Thanks for this - it seems to have partially fixed an issue I was having.
> 
> I have a dehumidifier, the humidity level settings are 15,35,40...... 80 - 15 is constant, then 35 -> 80 are actual humidity levels.
> 
> I managed to map all the various other DPs, but this one would not work until I used your modifications (select would show 'unavailable' and number just wouldn't do anything).
> 
> With the changes to number and the ability to specify 'int' I can now create a slider (with 5 value step) that controls the humidity level.
> 
> But, I think 'select' would be a better option (so that I can skip 20, 25, 30 values), whilst using select with your modifications now shows the values in HA, if I try to change them using the select entity the values do not make it to the device (changes from HA are ignored by the device, but it can read the state from the device)
> 
> ... long story short, I'm wondering if the CONF_DP_DTYPE = "dp_data_type" (so that int can be specified) added to the 'select' option might also be valuable and resolve the problem? I assume this would require changes to:
> 
> `async def async_select_option`
> 
> Rather than just:
> 
> `def status_updated(self):`
> 
> I will try hacking around with your changes to see.

I'm working on a more direct approach to this.
Based on all of the weird Tuya devices I have, I believe that each DPS can only hold 1 type of data (int, float, str).
So, what I did was implement a way for it to knows what each data type for each DP is and cast before sending it to the device.
However, this is a drastic change in behavior and need to be discussed with the owner of this repo.

Basically, at the first time the device is added, when adding each entity, it'll ask for the dtype which will default to the dtype that it thinks is correct. In order to prevent this to be a breaking change, we might even need to scan the dtype each time the connection is made, if not specify.

This really require discussion @rospogrigio what do you think?

@iMiMx However, this would only solve the root cause -- the data type -- only. We still need someone to add step to input number and cast the state value of select to str.

**Comment by Zen3515** [comment]
After two years of waiting, if anyone is still encountering this problem, best of luck! I don't care anymore.

How hard is it to change set_dp value type...

No comment, no response, nothing.
If the issue doesn't affect the repo owner personally, they're unlikely to show any interest or response to someone else's problem.

My best suggestion for you to disable that dp from local tuya and add your own number / select.
For example:

```
service: localtuya.set_dp
data:
  device_id: abcabcabcabcabc
  dp: 106
  value: 1
```
